src/Nethermind/Nethermind.Synchronization/Blocks/BlockDownloader.cs

```diff
@@ -80,91 +109,52 @@ private void BlockTreeOnNewHeadBlock(object? sender, BlockEventArgs e)
-                if (_logger.IsError) _logger.Error(message);
-                throw new ArgumentNullException(message);
-            }
-
-            DownloaderOptions options = blocksRequest.Options;
             bool originalDownloadReceiptOpts = (options & DownloaderOptions.WithReceipts) == DownloaderOptions.WithReceipts;
             bool originalShouldProcess = (options & DownloaderOptions.Process) == DownloaderOptions.Process;
 
             int blocksSynced = 0;
-            // pivot number - 6 for uncle validation
-            // long currentNumber = Math.Max(Math.Max(0, pivotNumber - 6), Math.Min(_blockTree.BestKnownNumber, bestPeer.HeadNumber - 1));
             long bestProcessedBlock = 0;
+            long previousStartingHeaderNumber = -1;
 
             while (true)
             {
-                using IOwnedReadOnlyList<BlockHeader?>? headers = await _forwardHeaderProvider.GetBlockHeaders(blocksRequest.NumberOfLatestBlocksToBeIgnored ?? 0, _syncBatchSize.Current, cancellation);
-                if (cancellation.IsCancellationRequested) return blocksSynced; // check before every heavy operation
-                if (headers is null || headers.Count <= 1) return blocksSynced;
-
-                (bool shouldProcess, bool downloadReceipts) = ReceiptEdgeCase(bestProcessedBlock, headers[0].Number, originalShouldProcess, originalDownloadReceiptOpts);
+                using IOwnedReadOnlyList<BlockHeader?>? headers = await _forwardHeaderProvider.GetBlockHeaders(fastSyncLag, HeaderLookupSize + 1, cancellation);
+                if (cancellation.IsCancellationRequested) return null; // check before every heavy operation
+                if (headers is null || headers.Count <= 1) return null;
 
-                BlockDownloadContext context = new(_specProvider, bestPeer, headers, downloadReceipts, _receiptsRecovery);
-                long startTime = Stopwatch.GetTimestamp();
-                await RequestBodies(bestPeer, cancellation, context);
-
-                if (context.DownloadReceipts)
+                if (previousStartingHeaderNumber == headers[0].Number)
                 {
-                    if (cancellation.IsCancellationRequested) return blocksSynced; // check before every heavy operation
-                    await RequestReceipts(bestPeer, cancellation, context);
+                    // Note: Could be change in peer or a fork.
+                    throw new InvalidOperationException("Forward header starting block number did not changed. This is unexpected");
                 }
+                previousStartingHeaderNumber = headers[0].Number;
 
-                AdjustSyncBatchSize(Stopwatch.GetElapsedTime(startTime));
-
-                Block[]? blocks = context.Blocks;
-                TxReceipt[]?[]? receipts = context.ReceiptsForBlocks;
+                (bool shouldProcess, bool downloadReceipts) = ReceiptEdgeCase(bestProcessedBlock, headers[1].Number, originalShouldProcess, originalDownloadReceiptOpts);
```

Comment:
Question about the edge case - if we processed some block, we shouldn't need to download receipt for it. If we store receipts that receipt should be already stored.

Reply:
If `shouldProcess` is originally true, then `shouldDownloadReceipt` is originally false (FullSync). Otherwise in FastSync, it download receipt depending on sync config. This is to detect edge case where FullSync is on but the state is not at just the right block, resulting in missing receipt. Yes, this does mean that all along the mode could just be `FullSync` all the time and it would probably be just fine.